What causes __bindgen_padding and how to avoid / deal with it when cross-compiling?

First of all, thanks for the great tool! It really makes it easy to bind against C code :-)

With my project I'm running into an issue and I'm not sure what is causing it. I'm integrating a C library into my project, which will run on embedded / ARM hardware. It (cross)compiles fine when using the `FROM rust:1.40-stretch` Docker image. Unfortunately I run into issues when I upgraded to `FROM rust:1.40-buster`. In the latter environment it complains that I'm not setting the `__bindgen_padding_` fields when initializing structs. What is different between these two environments is the clang and arm-linux-gnueabihf package versions. I have seen the same issue when using Yocto Zeus, meta-rust and meta-clang (which uses clang / llvm 9.0 I believe).

**I'm looking for some advise on how to deal with these `__bindgen_padding_` fields and how I can potentially avoid them being generated.**

## `rust:1.40-stretch` environment 

* clang 3.9
* gcc-arm-linux-gnueabihf 6.3
* g++-arm-linux-gnueabihf 6.3

### Native

Compiles fine.

### ARM cross-compiling

Compiles fine.

## `rust:1.40-buster` environment

* clang 7.0
* gcc-arm-linux-gnueabihf 8.3
* g++-arm-linux-gnueabihf 8.3

### Native

Compiles fine.

### ARM cross-compiling

I get the following errors:

```
   Compiling libloragw-sx1302 v1.0.0 (/chirpstack-concentratord/libloragw-sx1302)
error[E0063]: missing field `__bindgen_padding_0` in initializer of `wrapper::coord_s`
   --> libloragw-sx1302/src/gps/mod.rs:183:19
    |
183 |     let mut loc = wrapper::coord_s {
    |                   ^^^^^^^^^^^^^^^^ missing `__bindgen_padding_0`

error[E0063]: missing field `__bindgen_padding_0` in initializer of `wrapper::coord_s`
   --> libloragw-sx1302/src/gps/mod.rs:189:19
    |
189 |     let mut err = wrapper::coord_s {
    |                   ^^^^^^^^^^^^^^^^ missing `__bindgen_padding_0`

error[E0063]: missing field `__bindgen_padding_0` in initializer of `wrapper::coord_s`
   --> libloragw-sx1301/src/gps/mod.rs:183:19
    |
183 |     let mut loc = wrapper::coord_s {
    |                   ^^^^^^^^^^^^^^^^ missing `__bindgen_padding_0`

error[E0063]: missing field `__bindgen_padding_0` in initializer of `wrapper::coord_s`
   --> libloragw-sx1301/src/gps/mod.rs:189:19
    |
189 |     let mut err = wrapper::coord_s {
    |                   ^^^^^^^^^^^^^^^^ missing `__bindgen_padding_0`

error[E0063]: missing field `__bindgen_padding_0` in initializer of `wrapper::lgw_conf_rxif_s`
   --> libloragw-sx1301/src/hal/mod.rs:275:16
    |
275 |     let conf = wrapper::lgw_conf_rxif_s {
    |                ^^^^^^^^^^^^^^^^^^^^^^^^ missing `__bindgen_padding_0`

error[E0063]: missing fields `__bindgen_padding_0`, `__bindgen_padding_1` in initializer of `wrapper::lgw_conf_rxif_s`
   --> libloragw-sx1302/src/hal/mod.rs:353:20
    |
353 |     let mut conf = wrapper::lgw_conf_rxif_s {
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^ missing `__bindgen_padding_0`, `__bindgen_padding_1`



```

I do understand the error (I'm trying to initialize a struct, but I'm not initializing the `__bindgen_padding_` field). What I do not understand what is causing the `__bindgen_padding_` / why it is there. Especially because it seems to depend on the version of the clang / arm-linux-gnueabihf package versions, this can be problematic because the __bindgen_padding_` might be there or it might not.

## Input C headers

Example of one of the headers:

```c
struct coord_s {
    double  lat;    /*!> latitude [-90,90] (North +, South -) */
    double  lon;    /*!> longitude [-180,180] (East +, West -)*/
    short   alt;    /*!> altitude in meters (WGS 84 geoid ref.) */
};
```

https://github.com/Lora-net/sx1302_hal/blob/master/libloragw/inc/loragw_gps.h#L51

## Reproducing the issue

The issue can be reproduced using the Docker Compose development environment:
https://github.com/brocaar/chirpstack-concentratord/tree/test

```git clone https://github.com/brocaar/chirpstack-concentratord -b test```

### `rust:1.40-stretch`

```
docker-compose run --rm chirpstack-concentratord-stretch bash

# native (works)
make build-native-debug

# arm crosscompile (works)
make build-armv7-debug
```


### `rust:1.40-buster`

```
docker-compose run --rm chirpstack-concentratord-buster bash

# native (works)
make build-native-debug

# arm crosscompile (fails)
make build-armv7-debug
```

--- a/src/codegen/struct_layout.rs
+++ b/src/codegen/struct_layout.rs
@@ -258,6 +258,11 @@ impl<'a> StructLayoutTracker<'a> {
         }
 
         let padding_bytes = layout.size - self.latest_offset;
+        if padding_bytes == 0 {
+            return None;
+        }
+
+        let repr_align = self.ctx.options().rust_features().repr_align;
 
         // We always pad to get to the correct size if the struct is one of
         // those we can't align properly.
@@ -265,12 +270,10 @@ impl<'a> StructLayoutTracker<'a> {
         // Note that if the last field we saw was a bitfield, we may need to pad
         // regardless, because bitfields don't respect alignment as strictly as
         // other fields.
-        if padding_bytes > 0 &&
-            (padding_bytes >= layout.align ||
-                (self.last_field_was_bitfield &&
-                    padding_bytes >=
-                        self.latest_field_layout.unwrap().align) ||
-                layout.align > self.ctx.target_pointer_size())
+        if padding_bytes >= layout.align ||
+             (self.last_field_was_bitfield &&
+              padding_bytes >= self.latest_field_layout.unwrap().align) ||
+             (!repr_align && layout.align > self.ctx.target_pointer_size())
         {
             let layout = if self.is_packed {
                 Layout::new(padding_bytes, 1)

--- a/src/codegen/struct_layout.rs
+++ b/src/codegen/struct_layout.rs
@@ -9,6 +9,8 @@ use ir::ty::{Type, TypeKind};
 use proc_macro2::{self, Ident, Span};
 use std::cmp;
 
+const MAX_GUARANTEED_ALIGN: usize = 8;
+
 /// Trace the layout of struct.
 #[derive(Debug)]
 pub struct StructLayoutTracker<'a> {
@@ -168,10 +170,10 @@ impl<'a> StructLayoutTracker<'a> {
             // much we can do about it.
             if let Some(layout) = self.ctx.resolve_type(inner).layout(self.ctx)
             {
-                if layout.align > self.ctx.target_pointer_size() {
+                if layout.align > MAX_GUARANTEED_ALIGN {
                     field_layout.size =
                         align_to(layout.size, layout.align) * len;
-                    field_layout.align = self.ctx.target_pointer_size();
+                    field_layout.align = MAX_GUARANTEED_ALIGN;
                 }
             }
         }
@@ -191,7 +193,7 @@ impl<'a> StructLayoutTracker<'a> {
 
             // Otherwise the padding is useless.
             let need_padding = padding_bytes >= field_layout.align ||
-                field_layout.align > self.ctx.target_pointer_size();
+                field_layout.align > MAX_GUARANTEED_ALIGN;
 
             self.latest_offset += padding_bytes;
 
@@ -213,10 +215,7 @@ impl<'a> StructLayoutTracker<'a> {
             if need_padding && padding_bytes != 0 {
                 Some(Layout::new(
                     padding_bytes,
-                    cmp::min(
-                        field_layout.align,
-                        self.ctx.target_pointer_size(),
-                    ),
+                    cmp::min(field_layout.align, MAX_GUARANTEED_ALIGN),
                 ))
             } else {
                 None
@@ -271,14 +270,14 @@ impl<'a> StructLayoutTracker<'a> {
         // regardless, because bitfields don't respect alignment as strictly as
         // other fields.
         if padding_bytes >= layout.align ||
-             (self.last_field_was_bitfield &&
-              padding_bytes >= self.latest_field_layout.unwrap().align) ||
-             (!repr_align && layout.align > self.ctx.target_pointer_size())
+            (self.last_field_was_bitfield &&
+                padding_bytes >= self.latest_field_layout.unwrap().align) ||
+            (!repr_align && layout.align > MAX_GUARANTEED_ALIGN)
         {
             let layout = if self.is_packed {
                 Layout::new(padding_bytes, 1)
             } else if self.last_field_was_bitfield ||
-                layout.align > self.ctx.target_pointer_size()
+                layout.align > MAX_GUARANTEED_ALIGN
             {
                 // We've already given up on alignment here.
                 Layout::for_size(self.ctx, padding_bytes)
@@ -309,9 +308,9 @@ impl<'a> StructLayoutTracker<'a> {
             return false;
         }
 
-        // We can only generate up-to a word of alignment unless we support
+        // We can only generate up-to a 8-bytes of alignment unless we support
         // repr(align).
-        repr_align || layout.align <= self.ctx.target_pointer_size()
+        repr_align || layout.align <= MAX_GUARANTEED_ALIGN
     }
 
     fn padding_bytes(&self, layout: Layout) -> usize {

--- a/tests/expectations/tests/issue-1291.rs
+++ b/tests/expectations/tests/issue-1291.rs
@@ -26,7 +26,6 @@ pub struct RTCRay {
     pub geomID: ::std::os::raw::c_uint,
     pub primID: ::std::os::raw::c_uint,
     pub instID: ::std::os::raw::c_uint,
-    pub __bindgen_padding_0: [u32; 3usize],
 }
 #[test]
 fn bindgen_test_layout_RTCRay() {

--- a/tests/expectations/tests/issue-648-derive-debug-with-padding.rs
+++ b/tests/expectations/tests/issue-648-derive-debug-with-padding.rs
@@ -16,7 +16,6 @@
 #[derive(Copy, Clone)]
 pub struct NoDebug {
     pub c: ::std::os::raw::c_char,
-    pub __bindgen_padding_0: [u8; 63usize],
 }
 #[test]
 fn bindgen_test_layout_NoDebug() {
@@ -61,7 +60,6 @@ impl ::std::cmp::PartialEq for NoDebug {
 pub struct ShouldDeriveDebugButDoesNot {
     pub c: [::std::os::raw::c_char; 32usize],
     pub d: ::std::os::raw::c_char,
-    pub __bindgen_padding_0: [u8; 31usize],
 }
 #[test]
 fn bindgen_test_layout_ShouldDeriveDebugButDoesNot() {

--- a/tests/expectations/tests/layout_array.rs
+++ b/tests/expectations/tests/layout_array.rs
@@ -68,7 +68,6 @@ pub struct rte_mempool_ops {
     pub dequeue: rte_mempool_dequeue_t,
     ///< Get qty of available objs.
     pub get_count: rte_mempool_get_count,
-    pub __bindgen_padding_0: [u64; 7usize],
 }
 #[test]
 fn bindgen_test_layout_rte_mempool_ops() {

--- a/tests/expectations/tests/layout_array_too_long.rs
+++ b/tests/expectations/tests/layout_array_too_long.rs
@@ -163,7 +163,6 @@ pub struct ip_frag_pkt {
     pub last_idx: u32,
     ///< fragments
     pub frags: [ip_frag; 4usize],
-    pub __bindgen_padding_0: [u64; 6usize],
 }
 #[repr(C)]
 #[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]

--- a/tests/expectations/tests/layout_kni_mbuf.rs
+++ b/tests/expectations/tests/layout_kni_mbuf.rs
@@ -33,7 +33,6 @@ pub struct rte_kni_mbuf {
     pub pad3: [::std::os::raw::c_char; 8usize],
     pub pool: *mut ::std::os::raw::c_void,
     pub next: *mut ::std::os::raw::c_void,
-    pub __bindgen_padding_1: [u64; 5usize],
 }
 #[test]
 fn bindgen_test_layout_rte_kni_mbuf() {

--- a/tests/expectations/tests/layout_large_align_field.rs
+++ b/tests/expectations/tests/layout_large_align_field.rs
@@ -193,7 +193,6 @@ pub struct ip_frag_pkt {
     pub last_idx: u32,
     ///< fragments
     pub frags: [ip_frag; 4usize],
-    pub __bindgen_padding_0: [u64; 6usize],
 }
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
@@ -417,7 +416,6 @@ pub struct ip_frag_tbl_stat {
     pub fail_total: u64,
     ///< # of 'no space' add failures.
     pub fail_nospace: u64,
-    pub __bindgen_padding_0: [u64; 2usize],
 }
 #[test]
 fn bindgen_test_layout_ip_frag_tbl_stat() {

--- a/tests/expectations/tests/layout_mbuf.rs
+++ b/tests/expectations/tests/layout_mbuf.rs
@@ -178,7 +178,6 @@ pub struct rte_mbuf {
     pub priv_size: u16,
     /// Timesync flags for use with IEEE1588.
     pub timesync: u16,
-    pub __bindgen_padding_0: [u32; 7usize],
 }
 /// 16-bit Reference counter.
 /// It should only be accessed using the following functions:

--- a/tests/expectations/tests/libclang-5/call-conv-field.rs
+++ b/tests/expectations/tests/libclang-5/call-conv-field.rs
@@ -14,7 +14,6 @@ pub struct JNINativeInterface_ {
     pub GetVersion: ::std::option::Option<
         unsafe extern "stdcall" fn(env: *mut ::std::os::raw::c_void) -> ::std::os::raw::c_int,
     >,
-    pub __bindgen_padding_0: u32,
     pub __hack: ::std::os::raw::c_ulonglong,
 }
 #[test]

--- a/tests/expectations/tests/libclang-9/call-conv-field.rs
+++ b/tests/expectations/tests/libclang-9/call-conv-field.rs
@@ -12,9 +12,10 @@
 #[derive(Default, Copy, Clone)]
 pub struct JNINativeInterface_ {
     pub GetVersion: ::std::option::Option<
-        unsafe extern "stdcall" fn(env: *mut ::std::os::raw::c_void) -> ::std::os::raw::c_int,
+        unsafe extern "stdcall" fn(
+            env: *mut ::std::os::raw::c_void,
+        ) -> ::std::os::raw::c_int,
     >,
-    pub __bindgen_padding_0: u32,
     pub __hack: ::std::os::raw::c_ulonglong,
 }
 #[test]
@@ -30,7 +31,10 @@ fn bindgen_test_layout_JNINativeInterface_() {
         concat!("Alignment of ", stringify!(JNINativeInterface_))
     );
     assert_eq!(
-        unsafe { &(*(::std::ptr::null::<JNINativeInterface_>())).GetVersion as *const _ as usize },
+        unsafe {
+            &(*(::std::ptr::null::<JNINativeInterface_>())).GetVersion
+                as *const _ as usize
+        },
         0usize,
         concat!(
             "Offset of field: ",
@@ -40,7 +44,10 @@ fn bindgen_test_layout_JNINativeInterface_() {
         )
     );
     assert_eq!(
-        unsafe { &(*(::std::ptr::null::<JNINativeInterface_>())).__hack as *const _ as usize },
+        unsafe {
+            &(*(::std::ptr::null::<JNINativeInterface_>())).__hack as *const _
+                as usize
+        },
         8usize,
         concat!(
             "Offset of field: ",

--- a/tests/expectations/tests/unknown_attr.rs
+++ b/tests/expectations/tests/unknown_attr.rs
@@ -14,7 +14,6 @@ pub struct max_align_t {
     pub __clang_max_align_nonce1: ::std::os::raw::c_longlong,
     pub __bindgen_padding_0: u64,
     pub __clang_max_align_nonce2: ::std::os::raw::c_longlong,
-    pub __bindgen_padding_1: u64,
 }
 #[test]
 fn bindgen_test_layout_max_align_t() {

